packages/react-router/lib/router/router.ts

```diff
@@ -4839,153 +4847,225 @@ function isSameRoute(
-    // values?".  If not, it should be safe to update in place.
-    let routeUpdates: Record<string, any> = {};
-    for (let lazyRouteProperty in lazyRoute) {
-      let staticRouteValue =
-        routeToUpdate[lazyRouteProperty as keyof typeof routeToUpdate];
-
-      let isPropertyStaticallyDefined =
-        staticRouteValue !== undefined &&
-        // This property isn't static since it should always be updated based
-        // on the route updates
-        lazyRouteProperty !== "hasErrorBoundary";
+  let cache = lazyRoutePropertyCache.get(routeToUpdate);
+  if (!cache) {
+    cache = {};
+    lazyRoutePropertyCache.set(routeToUpdate, cache);
+  }
 
-      warning(
-        !isPropertyStaticallyDefined,
-        `Route "${routeToUpdate.id}" has a static property "${lazyRouteProperty}" ` +
-          `defined but its lazy function is also returning a value for this property. ` +
-          `The lazy route property "${lazyRouteProperty}" will be ignored.`
-      );
+  let cachedPromise = cache[key];
+  if (cachedPromise) {
+    return cachedPromise;
+  }
+
+  let propertyPromise = (async () => {
+    let isUnsupported = isUnsupportedLazyRouteObjectKey(key);
+    let staticRouteValue = routeToUpdate[key as keyof typeof routeToUpdate];
+    let isStaticallyDefined =
+      staticRouteValue !== undefined && key !== "hasErrorBoundary";
 
+    if (isUnsupported) {
       warning(
-        !unsupportedLazyRouteFunctionKeys.has(
-          lazyRouteProperty as UnsupportedLazyRouteFunctionKey
-        ),
+        !isUnsupported,
         "Route property " +
-          lazyRouteProperty +
-          " is not a supported property to be returned from a lazy route function. This property will be ignored."
+          key +
+          " is not a supported lazy route property. This property will be ignored."
+      );
+      cache[key] = Promise.resolve();
+    } else if (isStaticallyDefined) {
+      warning(
+        false,
+        `Route "${routeToUpdate.id}" has a static property "${key}" ` +
+          `defined. The lazy property will be ignored.`
       );
+    } else {
+      let value = await lazyFn();
+      if (value != null) {
+        Object.assign(routeToUpdate, { [key]: value });
+        Object.assign(routeToUpdate, mapRouteProperties(routeToUpdate));
+      }
+    }
 
+    // Clean up lazy property and clean up lazy object if it's now empty
+    if (typeof routeToUpdate.lazy === "object") {
+      routeToUpdate.lazy[key as keyof typeof routeToUpdate.lazy] = undefined;
       if (
-        !isPropertyStaticallyDefined &&
-        !unsupportedLazyRouteFunctionKeys.has(
-          lazyRouteProperty as UnsupportedLazyRouteFunctionKey
-        )
+        Object.values(routeToUpdate.lazy).every((value) => value === undefined)
       ) {
-        routeUpdates[lazyRouteProperty] =
-          lazyRoute[lazyRouteProperty as keyof typeof lazyRoute];
+        routeToUpdate.lazy = undefined;
       }
     }
```

Comment:
We perform this cleanup because the existing `route.lazy` logic uses the presence of a `route.lazy` value to decide whether to wait on a lazy call or not.